# Project 3 - Madelon Data Analysis
## Step 0: EDA

In [1]:
!conda install psycopg2 --yes

Fetching package metadata .............
Solving package specifications: .

# All requested packages already installed.
# packages in environment at /opt/conda:
#
psycopg2                  2.7.3.1          py36h369a60c_0    defaults


In [3]:
%run madelon_analyzer.py
%run pickle_jar.py
%matplotlib inline

In [3]:
bd_6506_sample.head()

_id  feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  \
0  112013  0.874623 -0.433386  1.282504  0.440211 -0.729979 -0.121770   
1  107302  1.122571  0.275028 -0.296846  0.752930 -0.473619 -0.819199   
2   10607 -0.246912 -0.004445  0.805970 -1.743573 -0.042794  1.096977   
3  150235 -0.382419 -0.304121  1.141382 -0.414432 -0.248223  1.065871   
4   90111  1.347762  0.746815  0.362632  0.846772 -1.262384  1.364104   

   feat_006  feat_007  feat_008   ...    feat_991  feat_992  feat_993  \
0 -0.756523 -0.311089 -1.047285   ...    1.954668  1.426505  0.396491   
1  0.022044 -2.374956  2.130340   ...    0.670175  0.168732 -1.674614   
2 -0.054843 -0.661225  0.473623   ...   -0.371504  0.905834 -1.494369   
3 -0.729584 -3.064909  0.430729   ...   -0.627376  1.258283 -0.653569   
4 -0.075601  0.352371 -1.001628   ...    0.431206  1.143164 -0.381148   

   feat_994  feat_995  feat_996  feat_997  feat_998  feat_999  target  
0 -0.705992 -0.312878  1.045555 -2.495671  1.039028  2.045024       0  
1  0.968680 -2.186223 -0.855602  0.118787 -0.765210 -0.077157       0  
2 -0.157062 -0.366484  0.746824  1.247538 -1.521214  1.343050       0  
3 -1.199824 -1.752736 -1.303956 -0.298844  0.772032  0.506764       1  
4  1.055597 -0.198674 -0.309390  0.105413  0.611066  0.372415       0  

[5 rows x 1002 columns]

# Big Data

Below is EDA of the expanded Madelon dataset with 1000 features and 200,000 observations. To start with, I've imported 3000 rows and will be performing unsupervised learning techniques to find and remove the noise.


In [8]:
def con_cur_to_class_db():
    '''This function will return a connection and cursor object from the expanded Madelon data that was provided 
    by the instructors. It is not the downloaded version of the data from the UCI website.'''
    
    con = pg2.connect(host='34.211.227.227',
                  dbname='postgres',
                  user='postgres')
    cur = con.cursor(cursor_factory=RealDictCursor)
    
    return con, cur

In [3]:
# in this cell, I'm going to try and get a representative sample of the population
# https://www.surveymonkey.com/mp/sample-size-calculator/ (confidence level = 90%, margin of error = 1)

# con_6506, cur_6506 = con_cur_to_class_db()
# %time cur_6506.execute('SELECT * FROM madelon ORDER BY random() LIMIT 6506')
# %time results_6506 = cur_6506.fetchall()
# con_6506.close()

In [ ]:
# # Dataframe of just 6506 samples for initial testing.

# samp_6506_bd_df = pd.DataFrame(results_6506)
# pd.to_pickle(samp_6506_bd_df, '6506_sample_big_data.p')

In [81]:
# samp_6506_bd_df.index = samp_6506_bd_df['_id']
# samp_6506_bd_df = samp_6506_bd_df.iloc[:,1:]

In [3]:
# con, cur = con_cur_to_class_db()
# %time cur.execute('SELECT * FROM madelon ORDER BY random() LIMIT 3000')
# %time results = cur.fetchall()
# con.close()

# # Dataframe of just 3000 samples for initial testing.

# samp_3000_bd_df = pd.DataFrame(results)

# pd.to_pickle(samp_3000_bd_df, '3000_sample_big_data.p')

In [9]:
samp_6506_bd_df.head()

_id  feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  \
0  112013  0.874623 -0.433386  1.282504  0.440211 -0.729979 -0.121770   
1  107302  1.122571  0.275028 -0.296846  0.752930 -0.473619 -0.819199   
2   10607 -0.246912 -0.004445  0.805970 -1.743573 -0.042794  1.096977   
3  150235 -0.382419 -0.304121  1.141382 -0.414432 -0.248223  1.065871   
4   90111  1.347762  0.746815  0.362632  0.846772 -1.262384  1.364104   

   feat_006  feat_007  feat_008   ...    feat_991  feat_992  feat_993  \
0 -0.756523 -0.311089 -1.047285   ...    1.954668  1.426505  0.396491   
1  0.022044 -2.374956  2.130340   ...    0.670175  0.168732 -1.674614   
2 -0.054843 -0.661225  0.473623   ...   -0.371504  0.905834 -1.494369   
3 -0.729584 -3.064909  0.430729   ...   -0.627376  1.258283 -0.653569   
4 -0.075601  0.352371 -1.001628   ...    0.431206  1.143164 -0.381148   

   feat_994  feat_995  feat_996  feat_997  feat_998  feat_999  target  
0 -0.705992 -0.312878  1.045555 -2.495671  1.039028  2.045024       0  
1  0.968680 -2.186223 -0.855602  0.118787 -0.765210 -0.077157       0  
2 -0.157062 -0.366484  0.746824  1.247538 -1.521214  1.343050       0  
3 -1.199824 -1.752736 -1.303956 -0.298844  0.772032  0.506764       1  
4  1.055597 -0.198674 -0.309390  0.105413  0.611066  0.372415       0  

[5 rows x 1002 columns]

In [10]:
X = samp_3000_bd_df.iloc[:200,1:-1]
X_2 = samp_3000_bd_df.iloc[200:400,1:-1]
X_3 = samp_3000_bd_df.iloc[400:600,1:-1]
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Columns: 1000 entries, feat_000 to feat_999
dtypes: float64(1000)
memory usage: 1.5 MB


In [11]:
# From this, I can tell that the big uci dataset is already normal
X.describe().T.mean()

count    200.000000
mean       0.001668
std        1.019437
min       -2.800271
25%       -0.681074
50%        0.001271
75%        0.682078
max        2.826687
dtype: float64

In [12]:
y = samp_3000_bd_df.iloc[:200,-1:]
y_2 = samp_3000_bd_df.iloc[200:400,-1:]
y_3 = samp_3000_bd_df.iloc[400:600,-1:]

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=42, test_size=0.333)
X_2_train, X_2_test, y_2_train, y_2_test = train_test_split(X_2,y_2, random_state=42, test_size=0.333)
X_3_train, X_3_test, y_3_train, y_3_test = train_test_split(X_3,y_3, random_state=42, test_size=0.333)

train_df = pd.merge(X_train, y_train, left_index=True, right_index=True)
train_2_df = pd.merge(X_2_train, y_2_train, left_index=True, right_index=True)
train_3_df = pd.merge(X_3_train, y_3_train, left_index=True, right_index=True)

test_df = pd.merge(X_test, y_test, left_index=True, right_index=True)
test_2_df = pd.merge(X_2_test, y_2_test, left_index=True, right_index=True)
test_3_df = pd.merge(X_3_test, y_3_test, left_index=True, right_index=True)

In [18]:
# NOTE: cell takes 7 minutes to run

ma = madelon_analyzer(train_df, test_df)

%time r2_bd_df = ma.mean_r2_for_all_features(train_df, DecisionTreeRegressor)

CPU times: user 6min 59s, sys: 160 ms, total: 6min 59s
Wall time: 6min 59s


In [23]:
r2_bd_df = r2_bd_df.sort_values('r2_score', ascending=False)

In [11]:
# pd.to_pickle(r2_bd_df, 'mean_r2_bd_sample_1.p')
# r2_bd_df = pd.read_pickle('Data/mean_r2_bd_sample_1.p')

In [ ]:
# ma_2 = madelon_analyzer(train_2_df, test_2_df)
# ma_3 = madelon_analyzer(train_3_df, test_3_df)

In [28]:
# %time r2_bd_2_df = ma_2.mean_r2_for_all_features(train_2_df, DecisionTreeRegressor)

CPU times: user 7min 6s, sys: 120 ms, total: 7min 6s
Wall time: 7min 6s


In [14]:
# r2_bd_2_df = r2_bd_2_df.sort_values('r2_score', ascending=False)
# pd.to_pickle(r2_bd_2_df, 'mean_r2_bd_sample_2.p')
r2_bd_2_df = pd.read_pickle('Data/mean_r2_bd_sample_2.p')

In [30]:
# %time r2_bd_3_df = ma_3.mean_r2_for_all_features(train_3_df, DecisionTreeRegressor)

CPU times: user 6min 48s, sys: 90 ms, total: 6min 48s
Wall time: 6min 48s


In [15]:
# r2_bd_3_df = r2_bd_3_df.sort_values('r2_score', ascending=False)
# pd.to_pickle(r2_bd_3_df, 'mean_r2_bd_sample_3.p')
r2_bd_3_df = pd.read_pickle('Data/mean_r2_bd_sample_3.p')

In [15]:
r2_bd_3_df[:20]

r2_score   feature
639   0.806627  feat_639
956   0.754375  feat_956
315   0.573237  feat_315
724   0.519356  feat_724
867   0.475584  feat_867
336   0.472721  feat_336
269   0.467021  feat_269
701   0.390734  feat_701
920   0.375489  feat_920
681   0.315068  feat_681
395   0.291377  feat_395
736   0.255419  feat_736
257   0.250702  feat_257
829    0.23143  feat_829
341   0.182425  feat_341
769   0.176203  feat_769
504   0.172482  feat_504
808   0.124137  feat_808
308 -0.0130461  feat_308
526  -0.329475  feat_526

In [16]:
r2_bd_2_df[:20]

r2_score   feature
639   0.845707  feat_639
956   0.808271  feat_956
315   0.575084  feat_315
867   0.485985  feat_867
341   0.462097  feat_341
701   0.417449  feat_701
724   0.409274  feat_724
526   0.347879  feat_526
769   0.327253  feat_769
829   0.314855  feat_829
269   0.311286  feat_269
336   0.293929  feat_336
920   0.290956  feat_920
395   0.284413  feat_395
257   0.279773  feat_257
808   0.211217  feat_808
736   0.200698  feat_736
504   0.198587  feat_504
681   0.188266  feat_681
308  0.0590268  feat_308

In [24]:
# First 20 features are all the same for the first 3 samples of the big madelon dataset

first_3_bd_samps_correlated_feats_list = \
list(set(list(r2_bd_df[:20].index)+list(r2_bd_2_df[:20].index)+list(r2_bd_3_df[:20].index)))

len(first_3_bd_samps_correlated_feats_list)

20

In [18]:
ma_first_3 = madelon_analyzer(samp_3000_bd_df[:2000], samp_3000_bd_df[2000:2600])
%time ma_first_3_benchmark = ma_first_3.train_val_scorer_df_maker(first_3_bd_samps_correlated_feats_list,\
                                                                  'first_3_bd')
ma_first_3_benchmark

CPU times: user 230 ms, sys: 40 ms, total: 270 ms
Wall time: 571 ms


first_3_bd
train_score    0.685500
val_score      0.523333

In [19]:
ma_first_3_6506 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
%time ma_first_3_benchmark_6506 = ma_first_3_6506.train_val_scorer_df_maker(first_3_bd_samps_correlated_feats_list,\
                                                                  'first_3_bd_6506')
ma_first_3_benchmark_6506

CPU times: user 1.47 s, sys: 20 ms, total: 1.49 s
Wall time: 3.3 s


first_3_bd_6506
train_score         0.678600
val_score           0.518592

In [ ]:
# ma_6506 = madelon_analyzer(samp_6506_bd_df.iloc[:200,500:], samp_6506_bd_df.iloc[500:600,500:])

# %time r2_bd_6506_df = ma_6506.mean_r2_for_all_features(samp_6506_bd_df.iloc[:500,:], DecisionTreeRegressor)

# r2_bd_6506_df = r2_bd_6506_df.sort_values('r2_score', ascending=False)

# pd.to_pickle(r2_bd_6506_df, 'mean_r2_bd_sample_6506.p')
# r2_bd_6506_df = pd.read_pickle('mean_r2_bd_sample_6506.p')

In [24]:
samp_6506_bd_df.head()

_id  feat_000  feat_001  feat_002  feat_003  feat_004  feat_005  \
0  112013  0.874623 -0.433386  1.282504  0.440211 -0.729979 -0.121770   
1  107302  1.122571  0.275028 -0.296846  0.752930 -0.473619 -0.819199   
2   10607 -0.246912 -0.004445  0.805970 -1.743573 -0.042794  1.096977   
3  150235 -0.382419 -0.304121  1.141382 -0.414432 -0.248223  1.065871   
4   90111  1.347762  0.746815  0.362632  0.846772 -1.262384  1.364104   

   feat_006  feat_007  feat_008   ...    feat_991  feat_992  feat_993  \
0 -0.756523 -0.311089 -1.047285   ...    1.954668  1.426505  0.396491   
1  0.022044 -2.374956  2.130340   ...    0.670175  0.168732 -1.674614   
2 -0.054843 -0.661225  0.473623   ...   -0.371504  0.905834 -1.494369   
3 -0.729584 -3.064909  0.430729   ...   -0.627376  1.258283 -0.653569   
4 -0.075601  0.352371 -1.001628   ...    0.431206  1.143164 -0.381148   

   feat_994  feat_995  feat_996  feat_997  feat_998  feat_999  target  
0 -0.705992 -0.312878  1.045555 -2.495671  1.039028  2.045024       0  
1  0.968680 -2.186223 -0.855602  0.118787 -0.765210 -0.077157       0  
2 -0.157062 -0.366484  0.746824  1.247538 -1.521214  1.343050       0  
3 -1.199824 -1.752736 -1.303956 -0.298844  0.772032  0.506764       1  
4  1.055597 -0.198674 -0.309390  0.105413  0.611066  0.372415       0  

[5 rows x 1002 columns]

In [25]:
X = samp_6506_bd_df.iloc[:,1:-1]
y = samp_6506_bd_df['target']

skb_20 = SelectKBest(k=20)
skb_20.fit(X, y)
skb_20_feats = np.where(skb_20.get_support())[0]
skb_20_cols = list(X[skb_20_feats].head().columns)

skb_20_cols_new = sorted([int(x[-3:]) for x in skb_20_cols])

In [26]:
pd.DataFrame([sorted(first_3_bd_samps_correlated_feats_list), skb_20_cols_new], \
             index=['first_3_bd_samps','skb_20']).T

first_3_bd_samps  skb_20
0                257     160
1                269     257
2                308     269
3                315     315
4                336     336
5                341     341
6                395     395
7                504     490
8                526     504
9                639     681
10               681     701
11               701     724
12               724     736
13               736     769
14               769     808
15               808     823
16               829     829
17               867     849
18               920     867
19               956     920

In [27]:
benchmarks_df = pd.merge(ma_first_3_benchmark, ma_first_3_benchmark_6506, left_index=True, right_index=True)

In [28]:
benchmarks_df

first_3_bd  first_3_bd_6506
train_score    0.685500         0.678600
val_score      0.523333         0.518592

In [29]:
ma_skb_6506 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_skb_benchmark_6506 = ma_skb_6506.train_val_scorer_df_maker(skb_20_cols_new,\
                                                                  'skb_20_feats')
ma_skb_benchmark_6506

skb_20_feats
train_score       0.67820
val_score         0.49668

In [30]:
skb_100 = SelectKBest(k=100)
skb_100.fit(X, y)
skb_100_feats = np.where(skb_100.get_support())[0]
skb_100_cols = list(X[skb_100_feats].head().columns)
skb_100_cols_new = sorted([int(x[-3:]) for x in skb_100_cols])
ma_skb_100 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_skb_100_scores = ma_skb_100.train_val_scorer_df_maker(skb_100_cols_new,\
                                                               'skb_100_feats')
ma_skb_100_scores

skb_100_feats
train_score       0.708800
val_score         0.552457

In [169]:
sfm = SelectFromModel(LassoCV(), threshold='mean')
sfm.fit(X, y)
sfm_feats = np.where(sfm.get_support())[0]
sfm_cols = list(X[sfm_feats].head().columns)
sfm_ma = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_sfm_scores = sfm_ma.train_val_scorer_df_maker(sfm_cols, 'sfm_feats')
ma_sfm_scores

sfm_feats
train_score   0.857000
val_score     0.762284

In [32]:
all_benchmark_feats = list(set((first_3_bd_samps_correlated_feats_list + skb_20_cols_new)))

In [33]:
# This shows that all 20 of the columns I selected are included in the top 100 seleced from 
benchmark_plus_skbtop100 = list(set(all_benchmark_feats + skb_100_cols_new))

In [146]:
ma_103 = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_103_scores = ma_103.train_val_scorer_df_maker(benchmark_plus_skbtop100, 'benchmark_plus_skbtop100')
ma_103_scores

benchmark_plus_skbtop100
train_score                  0.712200
val_score                    0.538513

In [4]:
feats_103 = sorted(['feat_{:3}'.format(x) for x in benchmark_plus_skbtop100])

NameError: name 'benchmark_plus_skbtop100' is not defined

In [10]:
feats_103 = \
'feat_004,feat_010,feat_061,feat_066,feat_078,feat_094,feat_100,feat_104,feat_107,feat_124,feat_139,feat_153,\
feat_154,feat_158,feat_160,feat_184,feat_188,feat_198,feat_229,feat_231,feat_245,feat_257,feat_263,feat_269,\
feat_276,feat_298,feat_304,feat_307,feat_308,feat_310,feat_314,feat_315,feat_316,feat_330,feat_336,feat_337,\
feat_341,feat_342,feat_344,feat_361,feat_368,feat_371,feat_380,feat_395,feat_396,feat_400,feat_413,feat_421,\
feat_426,feat_442,feat_449,feat_452,feat_470,feat_478,feat_487,feat_490,feat_496,feat_497,feat_504,feat_513,\
feat_526,feat_545,feat_553,feat_589,feat_595,feat_599,feat_638,feat_639,feat_660,feat_674,feat_681,feat_701,\
feat_710,feat_724,feat_736,feat_738,feat_769,feat_778,feat_781,feat_789,feat_808,feat_820,feat_823,feat_829,\
feat_830,feat_849,feat_862,feat_867,feat_885,feat_889,feat_895,feat_896,feat_898,feat_920,feat_941,feat_948,\
feat_954,feat_956,feat_969,feat_970,feat_977,feat_995,feat_996'

In [14]:
con_103, cur_103 = con_cur_to_class_db()
cur_103.execute("SELECT {}, target FROM madelon LIMIT 100000".format(feats_103))
results_103 = cur_103.fetchall()
con_103.close()

In [15]:
results_103_100000rows_df = pd.DataFrame(results_103)

In [16]:
results_103_100000rows_df.head()

feat_004  feat_010  feat_061  feat_066  feat_078  feat_094  feat_100  \
0 -0.313364 -1.853105 -0.617570  1.213093 -1.691096 -0.823690 -0.527816   
1 -0.252419  0.393065  0.146298  0.290831  1.717098  0.860437  0.548732   
2 -1.891341  0.562609  1.943417 -0.387551 -0.180743  2.008322  0.004553   
3 -2.458532 -0.589838 -0.372567  1.999407  0.855325 -0.335771  0.168373   
4  0.621189  2.448886  0.379100  0.581095  0.373658  0.007197 -1.526137   

   feat_104  feat_107  feat_124   ...    feat_941  feat_948  feat_954  \
0  0.261401 -1.291643  0.006729   ...   -2.186616  1.307990 -0.191312   
1  2.037863  0.933097 -0.994218   ...   -1.157805 -0.722995  1.135604   
2 -1.010499 -0.810443 -1.123838   ...    0.675879  0.635782  0.453698   
3 -0.173113 -0.989014  0.227036   ...   -0.081521 -0.965682 -0.319073   
4  0.484845 -1.108698 -0.640530   ...   -0.188028 -0.598852 -0.139465   

   feat_956  feat_969  feat_970  feat_977  feat_995  feat_996  target  
0  1.535919 -0.330030  1.385152  0.338486 -0.249347 -0.021732       0  
1 -0.672081  1.846824  0.874218 -0.924689  0.027623 -0.397976       0  
2  1.103762 -0.930680 -0.494260  1.046918  0.993074 -0.580654       0  
3 -0.574551  2.671121 -0.479038 -0.552304  0.977553  0.286620       1  
4  1.913404  0.216031 -0.209423  0.921488  0.773503 -1.985925       1  

[5 rows x 104 columns]

In [19]:
X_103 = results_103_all_rows_df.T[:-1].T
y_103 = results_103_all_rows_df.T[-1:].T

sfm_103 = SelectFromModel(LassoCV(), threshold='mean')
sfm_103.fit(X_103, y_103)
sfm_103_feats = np.where(sfm_103.get_support())[0]
sfm_103_cols = list(X_103.head(1).columns)
sfm_103_ma = madelon_analyzer(results_103_all_rows_df, samp_6506_bd_df)
ma_sfm_103_scores = sfm_103_ma.train_val_scorer_df_maker(sfm_103_cols, 'sfm103_feats_100000rows')
ma_sfm_103_scores

/opt/conda/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:1094: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


OSError: [Errno 28] No space left on device

In [113]:
sfm2 = SelectFromModel(LassoCV(), threshold='mean')
sfm2.fit(X[all_benchmark_feats], y)
sfm2_feats = np.where(sfm2.get_support())[0]
sfm2_cols = list(X[sfm2_feats].head().columns)
sfm2_ma = madelon_analyzer(samp_6506_bd_df[:5000], samp_6506_bd_df[5000:])
ma_sfm2_scores = sfm2_ma.train_val_scorer_df_maker(sfm2_cols, 'sfm2_feats')
ma_sfm2_scores

sfm2_feats
train_score    0.699000
val_score      0.508632

In [147]:
benchmarks_df = pd.merge(ma_first_3_benchmark, ma_first_3_benchmark_6506, left_index=True, right_index=True)
benchmarks_df = pd.merge(benchmarks_df, ma_skb_benchmark_6506, left_index=True, right_index=True)
benchmarks_df = pd.merge(benchmarks_df, ma_sfm_scores, left_index=True, right_index=True)
benchmarks_df = pd.merge(benchmarks_df, ma_sfm2_scores, left_index=True, right_index=True)
benchmarks_df = pd.merge(benchmarks_df, ma_skb_100_scores, left_index=True, right_index=True)
benchmarks_df = pd.merge(benchmarks_df, ma_103_scores, left_index=True, right_index=True)
multi_benchmarks_df = benchmarks_df.T.sort_values('val_score', ascending=False)
multi_benchmarks_df['description'] = ['Top 5 using SelectFromModel with LassoCV, threshold=mean', 
                                     'Top 100 using SelectKBest', 'SKB top 100 plus top 20',
                                      'Top 20 from the small samples', 
                                     'Using the same list from Top 20 from the small samples but on representitive sized sample',
                                     'SFM, on only the top 20 from small samples features', 'SKB top 20']
pd.to_pickle(multi_benchmarks_df, 'Data/multi_benchmarks_df.p')

In [148]:
pd.to_pickle(multi_benchmarks_df, 'Data/multi_benchmarks_df.p')
multi_benchmarks_df = pd.read_pickle('Data/multi_benchmarks_df.p')
multi_benchmarks_df

train_score  val_score  \
sfm_feats                      0.8570   0.762284   
skb_100_feats                  0.7088   0.552457   
benchmark_plus_skbtop100       0.7122   0.538513   
first_3_bd                     0.6855   0.523333   
first_3_bd_6506                0.6786   0.518592   
sfm2_feats                     0.6990   0.508632   
skb_20_feats                   0.6782   0.496680   

                                                                description  
sfm_feats                 Top 5 using SelectFromModel with LassoCV, thre...  
skb_100_feats                                     Top 100 using SelectKBest  
benchmark_plus_skbtop100                            SKB top 100 plus top 20  
first_3_bd                                    Top 20 from the small samples  
first_3_bd_6506           Using the same list from Top 20 from the small...  
sfm2_feats                SFM, on only the top 20 from small samples fea...  
skb_20_feats                                                     SKB top 20

## Thoughts after benchmarking and initial feature selection

Using a naive KNeighborsClassifier, I was able to score a 0.762284 for my validation set. From three random 1% samples of the data, I was able to identify 20 colinear features, which scored 0.518592 when I validated it on a random sample of 6506 rows, which is a representative sample of the population with a 90% confidence level and a 1% margin of error. 

Using SelectKBest to get 100 features, I was able to score higher on my validation set than when I looked for 20 features. This tells me that I was probably missing some of the important features in the set of 20 using this method. Although, I descovered that my benchmark list of 20 included 3 features that were not in the SKB top 100 list. So I added them, and will be using that full list.

Moving forward, I'm going to play with these lists to see if I can get a higher score than the 5 features from SelectFromModel.

---

## Feature Importance

In [153]:
val_df = samp_3000_bd_df.T[1:].T
train_df = samp_6506_bd_df.T[1:].T

In [165]:
rf_params = {
    'n_estimators':[10,50,100,200],
    'max_features':['auto','log2'],
}

ma_fi = madelon_analyzer(train_df, val_df)
%time ma_fi_scores_5 = ma_fi.train_val_scorer_df_maker(sfm_cols, 'SFM top 5 features, RFC',\
                                                     RandomForestClassifier, rf_params)
ma_fi_scores_5

CPU times: user 2.35 s, sys: 90 ms, total: 2.44 s
Wall time: 12.4 s


SFM top 5 features, RFC
train_score                 1.000000
val_score                   0.763333

In [170]:
rf_params = {
    'n_estimators':[10,50,100,200],
    'max_features':['auto','log2'],
}

ma_fi = madelon_analyzer(train_df, val_df)
%time ma_fi_scores_103 = ma_fi.train_val_scorer_df_maker(benchmark_plus_skbtop100, '103 features, RFC',\
                                                     RandomForestClassifier, rf_params)
ma_fi_scores_103

CPU times: user 11.7 s, sys: 110 ms, total: 11.8 s
Wall time: 41.1 s


103 features, RFC
train_score           1.000000
val_score             0.759667

In [ ]:
gbparams = {
    'learning_rate': [0.1,0.5,1], 
    'n_estimators': [100,200], 
    'max_depth': [2,4], 
}

ma_gb = madelon_analyzer(train_df, val_df)
%time ma_gb_scores_103 = ma_fi.train_val_scorer_df_maker(benchmark_plus_skbtop100, '103 features, GBC',\
                                                     GradientBoostingClassifier, gbparams)
ma_gb_scores_103

In [ ]:
p

--- 
 
---

### UCI Web Data

In [67]:
# Assigning file paths to variables. 

train_data_uci_madelon = './web_madelon_data/madelon_train.data.txt'
train_label_uci_madelon = './web_madelon_data/madelon_train.labels.txt'

val_data_uci_madelon = './web_madelon_data/madelon_valid.data.txt'
val_label_uci_madelon = './web_madelon_data/madelon_valid.labels.txt'

test_data_uci_madelon = './web_madelon_data/madelon_test.data.txt'

params_uci_madelon = './web_madelon_data/madelon.param.txt'

In [68]:
# Creating dataframes for the train, test, and val datasets.

train_uci_df = pd.read_csv(train_data_uci_madelon, delimiter=' ', header=None).drop(500, axis=1)
test_uci_df = pd.read_csv(test_data_uci_madelon, delimiter=' ', header=None).drop(500, axis=1)
val_uci_df = pd.read_csv(val_data_uci_madelon, delimiter=' ', header=None).drop(500, axis=1)

In [69]:
# Checking to make sure I have the right amount of rows.

print(len(val_uci_df) + len(train_uci_df) + len(test_uci_df))
print(len(val_uci_df))
print(len(train_uci_df))
print(len(test_uci_df))

4400
600
2000
1800


In [70]:
# Creating column names for all of the uci dataframes.

feature_col_names = ['feat_{}'.format(i) for i in range(0,500)]

train_uci_df.columns = feature_col_names
test_uci_df.columns = feature_col_names
val_uci_df.columns = feature_col_names

In [71]:
y_train = pd.read_csv(train_label_uci_madelon, header=None)
y_val = pd.read_csv(val_label_uci_madelon, header=None)

y_train.columns = ['target']
y_val.columns = ['target']

In [ ]:
train_uci_df = pd.merge(train_uci_df, y_train, left_index=True, right_index=True)
val_uci_df = pd.merge(val_uci_df, y_val, left_index=True, right_index=True)

In [ ]:
train_uci_df.head()

### Creating my three training sets that are 10% of the sample.

In [ ]:
s1_train_uci_10 = train_uci_df.sample(int(len(train_uci_df)*0.1), random_state=1)
s2_train_uci_10 = train_uci_df.sample(int(len(train_uci_df)*0.1), random_state=10)
s3_train_uci_10 = train_uci_df.sample(int(len(train_uci_df)*0.1), random_state=100)

In [ ]:
s1_train_uci_10.info()

In [ ]:
s1_dtree_mean_scores = []
for col in s1_train_uci_10.drop('target', axis=1).columns:
    s1_dtree_mean_scores.append(mean_r2_for_feature(s1_train_uci_10.drop('target', axis=1), col, DecisionTreeRegressor))

In [ ]:
s1_dtree_score_df = pd.DataFrame([s1_dtree_mean_scores, feature_col_names]).T.sort_values(0, ascending=False)
s1_dtree_score_df.head(25)

In [ ]:
s1_correlated_cols = list(s1_dtree_score_df.head(20).index)

In [ ]:
# In the above calculation, I have exactly 20 features that have positive scores. There is 5 predictors and 15 linear
    # combinations, so I think I've identified my noise! Next, I'm going to do this for the other random 10% samples
    # and compare the results. Maybe keep the top 20 of all three for my next step, using lass/ridge/elastic net to
    # remove features that are redundant.

In [ ]:
s2_dtree_mean_scores = []
for col in s2_train_uci_10.drop('target', axis=1).columns:
    s2_dtree_mean_scores.append(mean_r2_for_feature(s2_train_uci_10.drop('target', axis=1), col, DecisionTreeRegressor))

In [ ]:
s2_dtree_score_df = pd.DataFrame([s2_dtree_mean_scores, feature_col_names]).T.sort_values(0, ascending=False)

In [ ]:
s2_correlated_cols = list(s2_dtree_score_df.head(20).index)

In [ ]:
s3_dtree_mean_scores = []
for col in s3_train_uci_10.drop('target', axis=1).columns:
    s3_dtree_mean_scores.append(mean_r2_for_feature(s3_train_uci_10.drop('target', axis=1), col, DecisionTreeRegressor))

In [ ]:
s3_dtree_score_df = pd.DataFrame([s3_dtree_mean_scores, feature_col_names]).T.sort_values(0, ascending=False)

In [ ]:
s3_correlated_cols = list(s3_dtree_score_df.head(20).index)

In [ ]:
# Added all of my top 20 columns for the 3 random 10% samples to see if there was variation, and they all got the exact
    # same features. This is a good sign!

len(set(s1_correlated_cols+s2_correlated_cols+s3_correlated_cols))

In [ ]:
s1_correlated_cols

In [ ]:
s_all_results_df = pd.DataFrame([s1_correlated_cols, s2_correlated_cols, s3_correlated_cols]).T
s_all_results_df.columns = ['sample_1', 'sample_2', 'sample_3']
s_all_results_df.sort('sample_1')

In [ ]:
s1_corr = s1_train_uci_10[s1_correlated_cols].corr()
s2_corr = s2_train_uci_10[s2_correlated_cols].corr()
s3_corr = s3_train_uci_10[s3_correlated_cols].corr()

In [ ]:
s1_corr_desc = s1_corr.describe().T
s1_corr_desc['mean'] = abs(s1_corr_desc['mean'].values)
s1_corr_desc = s1_corr_desc.sort_values('mean')

In [ ]:
s2_corr_desc = s2_corr.describe().T
s2_corr_desc['mean'] = abs(s2_corr_desc['mean'].values)
s2_corr_desc = s2_corr_desc.sort_values('mean')

In [ ]:
s3_corr_desc = s1_corr.describe().T
s3_corr_desc['mean'] = abs(s3_corr_desc['mean'].values)
s3_corr_desc = s3_corr_desc.sort_values('mean')

In [ ]:
s3_skb_cols

In [ ]:
s3_corr_desc[['mean']][0:5]

In [ ]:
low_corr_cols = list(set(list(s1_corr_desc.head(7).index)+list(s2_corr_desc.head(7).index)+list(s3_corr_desc.head(7).index)))

In [ ]:
low_corr_corr = s2_train_uci_10[low_corr_cols].corr()

In [ ]:
# Compute the correlation matrix

corr = s1_corr

# Generate a mask for the upper triangle
mask = np.zeros_like(corr, dtype=np.bool)
mask[np.triu_indices_from(mask)] = True

# Set up the matplotlib figure
f, ax = plt.subplots(figsize=(11, 9))

# Generate a custom diverging colormap
cmap = sns.diverging_palette(220, 10, as_cmap=True)

# Draw the heatmap with the mask and correct aspect ratio
sns.heatmap(corr, mask=mask, cmap=cmap, vmax=.3, center=0,
            square=True, linewidths=.5, cbar_kws={"shrink": .5}, annot=True)

In [ ]:
# I realized this might not be a good method for finding redundant features. Leaving here in case I use it later.
s1_redundant_feats = list(set([433,153,128,472,28,451,318,281,433,475,241,28,451,48,378,453,493,64,336,128,105]))
len(s1_redundant_feats)

In [ ]:
# Getting the target values associated with each sample set.

s1_train_uci_10_with_y = train_uci_df.sample(int(len(train_uci_df)*0.1), random_state=1)

s2_train_uci_10_with_y = train_uci_df.sample(int(len(train_uci_df)*0.1), random_state=10)

s3_train_uci_10_with_y = train_uci_df.sample(int(len(train_uci_df)*0.1), random_state=100)

y_s1 = s1_train_uci_10_with_y[s1_correlated_cols+[-1]]['target']

y_s2 = s2_train_uci_10_with_y[s2_correlated_cols+[-1]]['target']

y_s3 = s3_train_uci_10_with_y[s3_correlated_cols+[-1]]['target']

from sklearn.feature_selection import SelectKBest 

# results f-classif: ['feat_241', 'feat_475', 'feat_64', 'feat_128', 'feat_336']
skb_s1 = SelectKBest(k=5)
skb_s1.fit(s1_train_uci_10[s1_correlated_cols], y_s1)
s1_skb_feats = np.where(skb_s1.get_support())[0]
s1_skb_cols = list(s1_train_uci_10[s1_correlated_cols][s1_skb_feats].head().columns)

s1_skb_cols

# results f-classif: ['feat_48', 'feat_475', 'feat_378', 'feat_241', 'feat_128']

skb_s2 = SelectKBest(k=5)
skb_s2.fit(s2_train_uci_10[s2_correlated_cols], y_s2)
s2_skb_feats = np.where(skb_s2.get_support())[0]
s2_skb_cols = list(s2_train_uci_10[s2_correlated_cols][s2_skb_feats].head().columns)

s2_skb_cols

# results f-classif: ['feat_241', 'feat_128', 'feat_378', 'feat_475', 'feat_338']

skb_s3 = SelectKBest(k=5)
skb_s3.fit(s3_train_uci_10[s3_correlated_cols], y_s3)
s3_skb_feats = np.where(skb_s3.get_support())[0]
s3_skb_cols = list(s3_train_uci_10[s3_correlated_cols][s3_skb_feats].head().columns)

s3_skb_cols

# Features that show in all three of the SelectKBest fits: 128, 241, 475
# Features that show twice: 378
# Features that show once: 64, 48, 338, 336

skb_all_samp_feats = set(s1_skb_cols + s2_skb_cols + s3_skb_cols)
skb_all_samp_feats

from sklearn.linear_model import Lasso, ElasticNet

# Found with Lasso: ['feat_442', 'feat_378', 'feat_475', 'feat_105', 'feat_336']
# Found with ElasticNet: ['feat_153','feat_442','feat_241','feat_378','feat_453','feat_105','feat_336','feat_455']
    
sfm_s1 = SelectFromModel(ElasticNet(), threshold='mean')
sfm_s1.fit(s1_train_uci_10[s1_correlated_cols], y_s1)
s1_sfm_feats = np.where(sfm_s1.get_support())[0]
s1_sfm_cols = list(s1_train_uci_10[s1_correlated_cols][s1_sfm_feats].head().columns)
s1_sfm_cols

# Found with Lasso: ['feat_475', 'feat_453', 'feat_64']
# Found with ElasticNet: ['feat_475', 'feat_336', 'feat_241', 'feat_64', 'feat_472']

sfm_s2 = SelectFromModel(ElasticNet(), threshold='mean')
sfm_s2.fit(s1_train_uci_10[s2_correlated_cols], y_s2)
s2_sfm_feats = np.where(sfm_s2.get_support())[0]
s2_sfm_cols = list(s2_train_uci_10[s2_correlated_cols][s2_sfm_feats].head().columns)
s2_sfm_cols

# Found with Lasso: ['feat_318', 'feat_378', 'feat_475', 'feat_338']
# Found with ElasticNet: ['feat_241','feat_318','feat_493','feat_378','feat_475','feat_455','feat_338']

sfm_s3 = SelectFromModel(ElasticNet(), threshold='mean')
sfm_s3.fit(s3_train_uci_10[s3_correlated_cols], y_s3)
s3_sfm_feats = np.where(sfm_s3.get_support())[0]
s3_sfm_cols = list(s3_train_uci_10[s3_correlated_cols][s3_sfm_feats].head().columns)
s3_sfm_cols

# Features that show in all three of the SelectFromModel fits: 475
# Features that show twice: 378
# Features that show once: 64, 105, 318, 336, 338, 442, 453

sfm_all_samp_feats = sorted(list(set(s1_sfm_cols + s2_sfm_cols + s3_sfm_cols)))
sfm_all_samp_feats

# Features that appear in both SFM and SKB: 64, 336, 338, 378, 475
# I feel like I should also consider 128 and 241 because that appeared in all three sample sets for SKB.

sfm_skb_all_feats_combo = set(list(skb_all_samp_feats) + list(sfm_all_samp_feats))
sfm_skb_all_feats_combo

# Initial thoughts:

#### UCI Web Data

Using the unsupervised learning technique where a regressor is used to compare every feature (column) against the rest of the dataset, I was able to find exactly 20 features that have mean R^2 scores that are positive, and most of them are very high (in the 0.8-0.9 range) for all three of the random data subsets. Since 480/500 of the features in the madelon dataset are noise, I believe that this is a good indication that I have found my 5 predictor features and the 15 linear combinations.

Using SelectKBest with k=5 on my 3 random data subsets, I was able to identify 8 unique features. Using SelectFromModel with the Lasso estimator, I was able to identify 9. Between these two lists of features, only 5 are the same. SelectFromModel returned 5 for the first dataset, 3 for the second, and 4 for the third. Since the full dataset has five important features, I think that the first sample is the most representative of the data. 

Moving forward, I will use GridSearchCV to tune my model hyperparameters and try various combonations of the important features I identified, starting with the 5 features that I found in both my SFM and SKB models. 